File: immoscout24-scraper/browser/python/immoscout24.py
"""Immoscout24 (Swiss) scraper using the official Scrapeless Python SDK + Playwright over CDP.
function names and emitted field names match verbatim.

Both surfaces extract from `window.__PINIA_INITIAL_STATE__`. the upstream reference returns
the nested JSON verbatim — Scrapeless mirrors it.
"""

from __future__ import annotations

import asyncio
import json
import os
from typing import Any, Dict, List

from loguru import logger
from parsel import Selector
from playwright.async_api import async_playwright
from scrapeless import Scrapeless
from scrapeless.types import ICreateBrowser

PROXY_COUNTRY = "CH"
DEFAULT_SESSION_TTL = 180

def _client() -> Scrapeless:
    if not os.environ.get("SCRAPELESS_API_KEY") and not os.environ.get("SCRAPELESS_KEY"):
        raise RuntimeError(
            "SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com"
        )
    if not os.environ.get("SCRAPELESS_API_KEY") and os.environ.get("SCRAPELESS_KEY"):
        os.environ["SCRAPELESS_API_KEY"] = os.environ["SCRAPELESS_KEY"]
    return Scrapeless()

async def _fetch_rendered_html(
    url: str,
    ready_selector: str,
    *,
    proxy_country: str = PROXY_COUNTRY,
    retries: int = 2,
    warmup: bool = True,
) -> str:
    """ImmoScout24 (SIX-fronted) interstitial-gates direct deep-links from
    cold sessions; a homepage warm-up gets the session cookie first.
    """
    last_error: Exception | None = None
    for attempt in range(retries + 1):
        client = _client()
        session = client.browser.create(
            ICreateBrowser(proxy_country=proxy_country, session_ttl=DEFAULT_SESSION_TTL)
        )
        ws = session.browser_ws_endpoint
        async with async_playwright() as pw:
            browser = await pw.chromium.connect_over_cdp(ws)
            try:
                page = await browser.new_page()
                if warmup:
                    try:
                        await page.goto(
                            "https://www.immoscout24.ch/",
                            wait_until="domcontentloaded",
                            timeout=30000,
                        )
                        await asyncio.sleep(3.5)
                    except Exception:
                        pass
                await page.goto(url, wait_until="domcontentloaded", timeout=60000)
                try:
                    await page.wait_for_selector(ready_selector, timeout=20000)
                except Exception as e:
                    logger.warning("wait_for_selector failed (continuing): {}", e)
                try:
                    await page.wait_for_function(
                        "window.__PINIA_INITIAL_STATE__ || document.body.innerText.length > 5000",
                        timeout=15000,
                    )
                except Exception:
                    pass
                html = await page.content()
                if html and len(html) > 20000:
                    return html
                last_error = RuntimeError("interstitial / short HTML")
            except Exception as e:  # noqa: BLE001
                last_error = e
                logger.warning("attempt {} failed: {}", attempt + 1, e)
            finally:
                try:
                    await browser.close()
                except Exception:
                    pass
    raise RuntimeError(f"giving up after {retries + 1} attempts: {last_error}")

# ---------------------------------------------------------------------------
# Pinia state extraction (bracket-depth matching — mirrors the upstream reference)
# ---------------------------------------------------------------------------

def _extract_pinia_state(html: str) -> Dict[str, Any] | None:
    sel = Selector(text=html)
    script_content = sel.xpath(
        "//script[contains(., 'window.__PINIA_INITIAL_STATE__')]/text()"
    ).get()
    if not script_content:
        return None
    idx = script_content.find("window.__PINIA_INITIAL_STATE__")
    start = script_content.find("{", idx)
    if start == -1:
        return None
    depth = 0
    end = start
    for i in range(start, len(script_content)):
        ch = script_content[i]
        if ch == "{":
            depth += 1
        elif ch == "}":
            depth -= 1
            if depth == 0:
                end = i + 1
                break
    json_str = script_content[start:end].replace("undefined", "null")
    try:
        return json.loads(json_str)
    except Exception as e:  # noqa: BLE001
        logger.warning("PINIA parse failed: {}", e)
        return None

# ---------------------------------------------------------------------------
# Parsers
# ---------------------------------------------------------------------------

def parse_property_page(html: str) -> Dict[str, Any]:
    data = _extract_pinia_state(html)
    if not data:
        return {}
    try:
        return data["listing"]["listing"]
    except (KeyError, TypeError):
        return {}

def parse_search_page(html: str) -> List[Dict[str, Any]]:
    """Each listing on the modern Immoscout24 SERP carries a schema.org
    `Product` JSON-LD block — the Pinia state isn't exposed on the window
    anymore. Fall back to JSON-LD when the legacy path returns nothing.
    """
    data = _extract_pinia_state(html)
    if data:
        try:
            legacy = data["resultList"]["search"]["fullSearch"]["result"]["listings"]
            if isinstance(legacy, list) and legacy:
                return legacy
        except (KeyError, TypeError):
            pass
    sel = Selector(text=html)
    out: List[Dict[str, Any]] = []
    for raw in sel.xpath('//script[@type="application/ld+json"]/text()').getall():
        try:
            node = json.loads(raw)
        except Exception:
            continue
        nodes = node if isinstance(node, list) else [node]
        for n in nodes:
            if not isinstance(n, dict):
                continue
            t = n.get("@type")
            types = t if isinstance(t, list) else [t]
            if "Product" not in types:
                continue
            name = str(n.get("name") or "")
            if "CHF" not in name.upper():
                continue
            offers = n.get("offers")
            offer = offers[0] if isinstance(offers, list) and offers else (offers if isinstance(offers, dict) else None)
            url = n.get("url")
            try:
                price = float(offer["price"]) if offer and "price" in offer else None
            except (TypeError, ValueError):
                price = None
            out.append({
                "id": url.rsplit("/", 1)[-1] if isinstance(url, str) else None,
                "name": name,
                "url": url,
                "image": n.get("image"),
                "description": n.get("description"),
                "price": price,
                "priceCurrency": offer.get("priceCurrency") if offer else None,
                "rawJsonLd": n,
            })
    return out

# ---------------------------------------------------------------------------
# Scrape functions — mirror the upstream reference's exports verbatim
# ---------------------------------------------------------------------------

async def scrape_properties(urls: List[str]) -> List[Dict[str, Any]]:
    out: List[Dict[str, Any]] = []
    for u in urls:
        html = await _fetch_rendered_html(
            u, ready_selector="script:has-text('__PINIA_INITIAL_STATE__')"
        )
        out.append(parse_property_page(html))
    return out

async def scrape_search(
    url: str,
    scrape_all_pages: bool = False,
    max_scrape_pages: int = 10,
) -> List[Dict[str, Any]]:
    first = await _fetch_rendered_html(
        url, ready_selector="script:has-text('__PINIA_INITIAL_STATE__')"
    )
    results = parse_search_page(first)

    # discover total pages
    data = _extract_pinia_state(first) or {}
    total_pages = 1
    try:
        total_pages = int(data["resultList"]["search"]["fullSearch"]["result"].get("numberOfPages", 1))
    except (KeyError, TypeError, ValueError):
        pass
    if not scrape_all_pages:
        total_pages = min(total_pages, max_scrape_pages)

    for page in range(2, total_pages + 1):
        sep = "&" if "?" in url else "?"
        page_url = f"{url}{sep}pn={page}"
        html = await _fetch_rendered_html(
            page_url, ready_selector="script:has-text('__PINIA_INITIAL_STATE__')"
        )
        page_items = parse_search_page(html)
        if not page_items:
            break
        results.extend(page_items)
    return results

def to_dict(obj: Any) -> Any:
    if isinstance(obj, list):
        return [to_dict(x) for x in obj]
    return obj

File: immoscout24-scraper/browser/nodejs/immoscout24.mjs
// Immoscout24 (Swiss) scraper using @scrapeless-ai/sdk + puppeteer-core over CDP.
//
// Mirrors  —
// function names and emitted field names match verbatim.

import { Scrapeless } from "@scrapeless-ai/sdk";
import puppeteer from "puppeteer-core";
import * as cheerio from "cheerio";

const PROXY_COUNTRY = "CH";
const DEFAULT_SESSION_TTL = 180;

function requireKey() {
  const k = process.env.SCRAPELESS_API_KEY ?? process.env.SCRAPELESS_KEY;
  if (!k) throw new Error("SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com");
  return k;
}

function client() {
  return new Scrapeless({ apiKey: requireKey() });
}

async function fetchRenderedHtml(url, readySelector, { proxyCountry = PROXY_COUNTRY, retries = 3, warmup = true } = {}) {
  let lastError;
  for (let attempt = 0; attempt <= retries; attempt++) {
    if (attempt > 0) await new Promise((r) => setTimeout(r, 8000 * attempt));
    const { browserWSEndpoint } = await client().browser.create({
      proxyCountry,
      sessionTTL: DEFAULT_SESSION_TTL,
    });
    let browser;
    try {
      browser = await puppeteer.connect({ browserWSEndpoint });
      const page = await browser.newPage();
      // ImmoScout24 (SIX-fronted, same anti-bot stack as Homegate) interstitial-
      // gates direct deep links. Warming up at the homepage gets a session
      // cookie so the next navigation returns the real Pinia state.
      if (warmup) {
        try {
          await page.goto("https://www.immoscout24.ch/", { waitUntil: "domcontentloaded", timeout: 30000 });
          await new Promise((r) => setTimeout(r, 3500));
        } catch (_) {}
      }
      await page.goto(url, { waitUntil: "domcontentloaded", timeout: 60000 });
      try { await page.waitForSelector(readySelector, { timeout: 20000 }); } catch (_) {}
      try {
        await page.waitForFunction(
          () => window.__PINIA_INITIAL_STATE__ || document.body.innerText.length > 5000,
          { timeout: 15000 },
        );
      } catch (_) {}
      const html = await page.content();
      if (html && html.length > 20000) return html;
      lastError = new Error("interstitial / short HTML");
    } catch (e) {
      lastError = e;
    } finally {
      try { await browser?.close(); } catch (_) {}
    }
  }
  throw new Error(`giving up after ${retries + 1} attempts: ${lastError?.message}`);
}

// Bracket-depth extraction of the Pinia state.
function extractPiniaState(html) {
  const $ = cheerio.load(html);
  let script = "";
  $("script").each((_, el) => {
    const txt = $(el).html() || "";
    if (txt.includes("window.__PINIA_INITIAL_STATE__")) {
      script = txt;
      return false;
    }
  });
  if (!script) return null;
  const idx = script.indexOf("window.__PINIA_INITIAL_STATE__");
  const start = script.indexOf("{", idx);
  if (start === -1) return null;
  let depth = 0;
  let end = start;
  for (let i = start; i < script.length; i++) {
    const ch = script[i];
    if (ch === "{") depth++;
    else if (ch === "}") {
      depth--;
      if (depth === 0) { end = i + 1; break; }
    }
  }
  const jsonStr = script.slice(start, end).replace(/undefined/g, "null");
  try { return JSON.parse(jsonStr); } catch (_) { return null; }
}

// ---------------- parsers ----------------

export function parsePropertyPage(html) {
  const data = extractPiniaState(html);
  if (!data) return {};
  return data?.listing?.listing ?? {};
}

// Each listing on the modern Immoscout24 SERP carries a schema.org `Product`
// JSON-LD block (name="<address> - CHF <price> - <rooms>", image, offers, url).
// Pinia state isn't exposed on the window anymore.
export function parseSearchPage(html) {
  const $ = cheerio.load(html);
  // First try the legacy Pinia path; if it lands data, prefer it.
  const data = extractPiniaState(html);
  const legacy = data?.resultList?.search?.fullSearch?.result?.listings;
  if (Array.isArray(legacy) && legacy.length) return legacy;
  const out = [];
  $('script[type="application/ld+json"]').each((_, el) => {
    let node;
    try { node = JSON.parse($(el).contents().text()); } catch (_) { return; }
    const nodes = Array.isArray(node) ? node : [node];
    for (const n of nodes) {
      if (!n || typeof n !== "object") continue;
      const t = n["@type"];
      const types = Array.isArray(t) ? t : [t];
      if (!types.includes("Product")) continue;
      const name = (n.name ?? "").toString();
      if (!/CHF/i.test(name)) continue; // skip the search-summary Product node
      const offer = Array.isArray(n.offers) ? n.offers[0] : n.offers;
      const priceText = offer?.price ?? null;
      const url = n.url ?? null;
      // DATA_MODEL.SearchResultEntry requires a `listing` sub-object so the
      // schema lines up with the legacy Pinia payload shape.
      out.push({
        listing: {
          id: url ? url.split("/").pop() : null,
          name,
          url,
          image: n.image ?? null,
          description: n.description ?? null,
          price: priceText ? parseFloat(priceText) : null,
          priceCurrency: offer?.priceCurrency ?? null,
          rawJsonLd: n,
        },
      });
    }
  });
  return out;
}

// ---------------- scrape functions ----------------

const PINIA_READY = "script";

export async function scrapeProperties(urls) {
  const out = [];
  for (const u of urls) {
    try {
      const html = await fetchRenderedHtml(u, PINIA_READY);
      const parsed = parsePropertyPage(html);
      // Skip empty objects (listing 404'd or page didn't render the Pinia state).
      if (parsed && Object.keys(parsed).length > 0) out.push(parsed);
    } catch (_) {
      // network error / interstitial — skip this URL
    }
  }
  return out;
}

export async function scrapeSearch(url, scrapeAllPages = false, maxScrapePages = 10) {
  const first = await fetchRenderedHtml(url, PINIA_READY);
  const results = parseSearchPage(first);
  const state = extractPiniaState(first) || {};
  let totalPages = 1;
  try {
    totalPages = parseInt(state?.resultList?.search?.fullSearch?.result?.numberOfPages ?? 1, 10) || 1;
  } catch (_) {}
  if (!scrapeAllPages) totalPages = Math.min(totalPages, maxScrapePages);

  for (let page = 2; page <= totalPages; page++) {
    const sep = url.includes("?") ? "&" : "?";
    const pageUrl = `${url}${sep}pn=${page}`;
    const html = await fetchRenderedHtml(pageUrl, PINIA_READY);
    const items = parseSearchPage(html);
    if (!items.length) break;
    results.push(...items);
  }
  return results;
}
